File: Go/040. Combination Sum II.go
package main

import "sort"

func combinationSum2(candidates []int, target int) [][]int {
	sort.Ints(candidates)
	return findSum(candidates, target)
}

func findSum(candidates []int, target int) [][]int {
	res := make([][]int, 0, 0)
	if len(candidates) == 0 || target < candidates[0] {
		return res
	}
	loc := 1
	for ; loc < len(candidates) && candidates[loc] == candidates[loc - 1]; loc++ {}
	curCandidates, leftCandidates := make([]int, 0), candidates[loc :]
	for i, num := 0, 0; i <= loc; i++ {
		if target - num == 0 {
			res = append(res, curCandidates)
			break
		} else if target - num > 0 {
			tmp := findSum(leftCandidates, target - num)
			for j := 0; j < len(tmp); j++ {
				tmp[j] = append(tmp[j], curCandidates...)
			}
			res = append(res, tmp...)
		} else {
			break
		}
		num += candidates[0]
		curCandidates = append(curCandidates, candidates[0])
	}
	return res
}

File: C++/040. Combination Sum II.cpp
#include <iostream>
#include <algorithm>
#include <vector>

class Solution {
public:
    std::vector<std::vector<int>> combinationSum2(std::vector<int>& candidates, int target)  {
        sort(candidates.begin(), candidates.end());
        return findSum(candidates, target);
    }

    std::vector<std::vector<int>> findSum(std::vector<int> candidates, int target) {
        std::vector<std::vector<int>> res;
        if (int(candidates.size()) == 0 || target < candidates[0])
            return res;
        int loc = 1;
        for (; loc < int(candidates.size()) && candidates[loc] == candidates[loc - 1]; loc++);
        std::vector<int> cur_candidates;
        std::vector<int> left_candidates(candidates.begin() + loc, candidates.end());
        for (int i = 0, num = 0; i <= loc; i++, num += candidates[0]) {
            if (target - num == 0) {
                res.emplace_back(cur_candidates);
            } else if (target - num > 0) {
                std::vector<std::vector<int>> tmp = findSum(left_candidates, target - num);
                for (auto &array : tmp)
                    array.insert(array.begin(), cur_candidates.begin(), cur_candidates.end());
                res.insert(res.end(), tmp.begin(), tmp.end());
            } else {
                break;
            }
            cur_candidates.emplace_back(candidates[0]);
        }
        return res;
    }
};